File: src/main/java/com/kelsos/mbrc/messaging/NotificationService.java
package com.kelsos.mbrc.messaging;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Build;
import android.support.v4.app.NotificationCompat;
import android.widget.RemoteViews;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.enums.PlayState;
import com.kelsos.mbrc.events.ui.NotificationDataAvailable;
import com.kelsos.mbrc.ui.activities.MainFragmentActivity;
import com.kelsos.mbrc.ui.activities.UpdateView;
import com.kelsos.mbrc.utilities.MainThreadBusWrapper;
import com.squareup.otto.Subscribe;

@Singleton
public class NotificationService {
    public static final int PLUGIN_OUT_OF_DATE = 15612;
    public static final int NOW_PLAYING_PLACEHOLDER = 15613;
    public static final String NOTIFICATION_PLAY_PRESSED = "com.kelsos.mbrc.notification.play";
    public static final String NOTIFICATION_NEXT_PRESSED = "com.kelsos.mbrc.notification.next";
    public static final String NOTIFICATION_CLOSE_PRESSED = "com.kelsos.mbrc.notification.close";
    private Context mContext;
    private MainThreadBusWrapper bus;

    @Inject public NotificationService(Context context, MainThreadBusWrapper bus) {
        this.mContext = context;
        this.bus = bus;
        bus.register(this);
    }

    @Subscribe public void handleNotificationData(NotificationDataAvailable event){
        notificationBuilder(event.getTitle(),event.getArtist(),event.getCover(),event.getState());
    }

    /**
     * Creates an ongoing notification that displays the cover and information about the playing track,
     * and also provides controls to skip or play/pause.
     * @param title The title of the track playing.
     * @param artist The artist of the track playing.
     * @param cover The cover Bitmap.
     * @param state The current play state is used to display the proper play or pause icon.
     */
    private void notificationBuilder(String title, String artist, Bitmap cover, PlayState state) {

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB) {
            return;
        }

        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(mContext);

        Intent notificationIntent = new Intent(mContext, MainFragmentActivity.class);
        PendingIntent notificationPendingIntent = PendingIntent.getActivity(mContext, 0, notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        mBuilder.setContentIntent(notificationPendingIntent);

        RemoteViews views = new RemoteViews(mContext.getPackageName(), R.layout.ui_notification_control);
        Intent playPressedIntent = new Intent(NOTIFICATION_PLAY_PRESSED);
        PendingIntent mediaPendingIntent = PendingIntent.getBroadcast(mContext, 1, playPressedIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        views.setOnClickPendingIntent(R.id.notification_play, mediaPendingIntent);
        Intent mediaNextButtonIntent = new Intent(NOTIFICATION_NEXT_PRESSED);
        PendingIntent mediaNextButtonPendingIntent = PendingIntent.getBroadcast(mContext, 2, mediaNextButtonIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        views.setOnClickPendingIntent(R.id.notification_next, mediaNextButtonPendingIntent);
        Intent clearNotificationIntent = new Intent(NOTIFICATION_CLOSE_PRESSED);
        PendingIntent clearNotificationPendingIntent = PendingIntent.getBroadcast(mContext, 3, clearNotificationIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        views.setOnClickPendingIntent(R.id.notification_close, clearNotificationPendingIntent);

        views.setTextViewText(R.id.notification_artist, artist);
        views.setTextViewText(R.id.notification_title, title);
        if (cover != null) {
            views.setImageViewBitmap(R.id.notification_album_art, cover);
        } else {
            views.setImageViewResource(R.id.notification_album_art, R.drawable.ic_image_no_cover);
        }

        switch (state) {
            case Playing:
                views.setImageViewResource(R.id.notification_play, R.drawable.ic_action_play);
                break;
            case Paused:
                views.setImageViewResource(R.id.notification_play, R.drawable.ic_action_pause);
                break;
            case Stopped:
                views.setImageViewResource(R.id.notification_play, R.drawable.ic_action_play);
                break;
            case Undefined:
                break;
        }

        views.setImageViewResource(R.id.notification_next, R.drawable.ic_action_next);

        views.setImageViewResource(R.id.notification_close, R.drawable.ic_action_collapse);


        Notification notification = mBuilder.build();
        notification.contentView = views;
        notification.flags = Notification.FLAG_ONGOING_EVENT;
        notification.icon = R.drawable.ic_mbrc_status;
        NotificationManager mNotificationManager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
        mNotificationManager.notify(NOW_PLAYING_PLACEHOLDER, notification);
    }

    public void cancelNotification(int notificationId) {
        NotificationManager mNotificationManager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
        mNotificationManager.cancel(notificationId);
    }

    public void updateAvailableNotificationBuilder() {
        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(mContext)
                .setSmallIcon(R.drawable.ic_mbrc_status)
                .setContentTitle(mContext.getString(R.string.application_name))
                .setContentText(mContext.getString(R.string.notification_plugin_out_of_date));

        Intent resultIntent = new Intent(mContext, UpdateView.class);
        resultIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);

        PendingIntent resultPendingIntent = PendingIntent.getActivity(mContext, 0, resultIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        mBuilder.setContentIntent(resultPendingIntent);
        final Notification notification = mBuilder.build();
        notification.flags = Notification.FLAG_AUTO_CANCEL;
        NotificationManager mNotificationManager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
        mNotificationManager.notify(PLUGIN_OUT_OF_DATE, notification);
    }
}

File: app/src/main/kotlin/com/kelsos/mbrc/messaging/NotificationService.kt
package com.kelsos.mbrc.messaging

import android.app.Application
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.graphics.BitmapFactory
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationCompat.Action
import androidx.core.app.NotificationManagerCompat
import com.kelsos.mbrc.R
import com.kelsos.mbrc.annotations.Connection
import com.kelsos.mbrc.annotations.PlayerState
import com.kelsos.mbrc.controller.ForegroundHooks
import com.kelsos.mbrc.events.bus.RxBus
import com.kelsos.mbrc.events.ui.ConnectionStatusChangeEvent
import com.kelsos.mbrc.events.ui.CoverChangedEvent
import com.kelsos.mbrc.events.ui.PlayStateChange
import com.kelsos.mbrc.events.ui.TrackInfoChangeEvent
import com.kelsos.mbrc.model.NotificationModel
import com.kelsos.mbrc.services.RemoteSessionManager
import com.kelsos.mbrc.utilities.RemoteUtils
import com.kelsos.mbrc.utilities.RemoteViewIntentBuilder.NEXT
import com.kelsos.mbrc.utilities.RemoteViewIntentBuilder.OPEN
import com.kelsos.mbrc.utilities.RemoteViewIntentBuilder.PLAY
import com.kelsos.mbrc.utilities.RemoteViewIntentBuilder.PREVIOUS
import com.kelsos.mbrc.utilities.RemoteViewIntentBuilder.getPendingIntent
import com.kelsos.mbrc.utilities.SettingsManager
import timber.log.Timber
import java.io.File
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class NotificationService
@Inject
constructor(
  bus: RxBus,
  private val context: Application,
  private val sessionManager: RemoteSessionManager,
  private val settings: SettingsManager,
  private val model: NotificationModel,
  private val notificationManager: NotificationManagerCompat
) {
  private var notification: Notification? = null
  private val previous: String
  private val play: String
  private val next: String
  private var hooks: ForegroundHooks? = null
  private val CHANNEL_ID = "mbrc_session"

  init {
    bus.register(this, TrackInfoChangeEvent::class.java) { this.handleTrackInfo(it) }
    bus.register(this, CoverChangedEvent::class.java) { this.coverChanged(it.path) }
    bus.register(this, PlayStateChange::class.java) { this.playStateChanged(it) }
    bus.register(this, ConnectionStatusChangeEvent::class.java) { this.connectionChanged(it) }
    previous = context.getString(R.string.notification_action_previous)
    play = context.getString(R.string.notification_action_play)
    next = context.getString(R.string.notification_action_next)
    createNotificationChannels()
  }

  private fun createNotificationChannels() {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
      return
    }

    val channel = NotificationChannel(
      CHANNEL_ID,
      "MusicBee Remote",
      NotificationManager.IMPORTANCE_DEFAULT
    )

    channel.apply {
      this.description = "MusicBee Remote: Service"
      enableLights(false)
      enableVibration(false)
      setSound(null, null)
    }

    notificationManager.createNotificationChannel(channel)
  }

  private fun handleTrackInfo(event: TrackInfoChangeEvent) {
    model.trackInfo = event.trackInfo
    update()
  }

  private fun coverChanged(path: String) {
    val coverFile = File(path)
    if (coverFile.exists()) {
      RemoteUtils.bitmapFromFile(coverFile.absolutePath).doOnTerminate {
        update()
      }.subscribe({
        model.cover = it
      }, {
        Timber.v(it, "failed to decode")
        model.cover = null
      })
    } else {
      model.cover = null
      update()
    }
  }

  private fun playStateChanged(event: PlayStateChange) {
    model.playState = event.state
    update()
  }

  private fun update() {
    notification = createBuilder().build().also { notification ->
      notificationManager.notify(NOW_PLAYING_PLACEHOLDER, notification)
    }
  }

  private fun connectionChanged(event: ConnectionStatusChangeEvent) {

    if (event.status == Connection.OFF) {
      cancelNotification(NOW_PLAYING_PLACEHOLDER)
    } else {
      notification = createBuilder().build()
      if (hooks != null) {
        hooks!!.start(NOW_PLAYING_PLACEHOLDER, notification!!)
      }
    }
  }

  private fun createBuilder(): NotificationCompat.Builder {
    val mediaStyle = androidx.media.app.NotificationCompat.MediaStyle()
    mediaStyle.setMediaSession(sessionManager.mediaSessionToken)

    val builder = NotificationCompat.Builder(context, CHANNEL_ID)
    val resId =
      if (model.playState == PlayerState.PLAYING) R.drawable.ic_action_pause else R.drawable.ic_action_play

    builder.setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
      .setSmallIcon(R.drawable.ic_mbrc_status)
      .setStyle(mediaStyle.setShowActionsInCompactView(1, 2))
      .addAction(previousAction)
      .addAction(getPlayAction(resId))
      .addAction(nextAction)

    builder.priority = NotificationCompat.PRIORITY_LOW
    builder.setOnlyAlertOnce(true)

    if (model.cover != null) {
      builder.setLargeIcon(model.cover)
    } else {
      val icon = BitmapFactory.decodeResource(context.resources, R.drawable.ic_image_no_cover)
      builder.setLargeIcon(icon)
    }

    val info = model.trackInfo

    if (info != null) {
      builder.setContentTitle(info.title).setContentText(info.artist).setSubText(info.album)
    }

    builder.setContentIntent(getPendingIntent(OPEN, context))

    return builder
  }

  private val previousAction: Action
    get() {
      val previousIntent = getPendingIntent(PREVIOUS, context)
      return Action.Builder(R.drawable.ic_action_previous, previous, previousIntent).build()
    }

  private fun getPlayAction(playStateIcon: Int): Action {
    val playIntent = getPendingIntent(PLAY, context)

    return Action.Builder(playStateIcon, play, playIntent).build()
  }

  private val nextAction: Action
    get() {
      val nextIntent = getPendingIntent(NEXT, context)
      return Action.Builder(R.drawable.ic_action_next, next, nextIntent).build()
    }

  fun cancelNotification(notificationId: Int) {
    notificationManager.cancel(notificationId)
    if (hooks != null) {
      hooks!!.stop()
    }
  }

  fun setForegroundHooks(hooks: ForegroundHooks) {
    this.hooks = hooks
    update()
    hooks.start(NOW_PLAYING_PLACEHOLDER, notification!!)
  }

  companion object {
    const val NOW_PLAYING_PLACEHOLDER = 15613
  }
}
